source/abstract/symbols.ts

```diff
@@ -0,0 +1,18 @@
+
+export const typeSymbol = 'type';
+export const tagSymbol = 'tag';
+
+// export type hasType<A extends string> = {type: A};
+// export type hasTag<A extends string> = {tag: A};
+// export type hasTypeAndTag<A extends string = string, B extends string = string> = hasType<A> & hasTag<B>;
+
+export abstract class HasTypeAndTag {
+    static get [typeSymbol](): string {
```

Comment:
It would be ideal for these to be abstract static properties, but typescript does not allow that (nor does it allow abstract static method); so instead this throws a not implemented error when called.
It would also be preferable for this to be a property rather than a method, but typescript only allows external symbols to be used in the definition of methods (but not properties); hence we use a method with a `get` prefix.